File: src/scheduler.js
import { monitor } from './index.js';

/**
 * Scheduler para execução periódica do monitoramento
 * Este script executa o monitoramento em intervalos regulares
 */

const INTERVAL_MINUTES = 60; // Intervalo padrão: 1 hora
const INTERVAL_MS = INTERVAL_MINUTES * 60 * 1000;

let isRunning = false;
let executionCount = 0;

/**
 * Executa o monitoramento
 */
async function runMonitoring() {
  if (isRunning) {
    console.log('⚠️  Monitoramento anterior ainda em execução, pulando...');
    return;
  }

  isRunning = true;
  executionCount++;

  console.log(`\n${'='.repeat(80)}`);
  console.log(`Execução #${executionCount} - ${new Date().toLocaleString('pt-BR')}`);
  console.log('='.repeat(80));

  try {
    const result = await monitor();
    
    if (result.status === 'saved') {
      console.log('✓ Nova observação coletada e salva!');
    } else if (result.status === 'duplicate') {
      console.log('ℹ️  Nenhuma nova observação disponível');
    } else if (result.status === 'error') {
      console.error('✗ Erro durante a execução:', result.error);
    }
  } catch (error) {
    console.error('✗ Erro inesperado:', error.message);
  } finally {
    isRunning = false;
  }

  console.log(`Próxima execução em ${INTERVAL_MINUTES} minutos...`);
}

/**
 * Inicia o scheduler
 */
function startScheduler() {
  console.log('🚀 T CrB Monitoring Scheduler Iniciado');
  console.log(`⏰ Intervalo de execução: ${INTERVAL_MINUTES} minutos`);
  console.log(`📅 Iniciado em: ${new Date().toLocaleString('pt-BR')}`);
  console.log('\nPressione Ctrl+C para parar\n');

  // Executa imediatamente na primeira vez
  runMonitoring();

  // Agenda execuções periódicas
  const intervalId = setInterval(runMonitoring, INTERVAL_MS);

  // Tratamento de sinais para encerramento gracioso
  process.on('SIGINT', () => {
    console.log('\n\n🛑 Encerrando scheduler...');
    clearInterval(intervalId);
    console.log(`✓ Scheduler encerrado após ${executionCount} execuções`);
    process.exit(0);
  });

  process.on('SIGTERM', () => {
    console.log('\n\n🛑 Encerrando scheduler...');
    clearInterval(intervalId);
    console.log(`✓ Scheduler encerrado após ${executionCount} execuções`);
    process.exit(0);
  });
}

// Inicia o scheduler se for executado diretamente
if (import.meta.url === `file://${process.argv[1]}`) {
  startScheduler();
}

export { startScheduler, runMonitoring };

File: scheduler.py
"""
Scheduler para execução periódica do monitoramento
Executa o monitoramento a cada hora
"""

import schedule
import time
import sys
from datetime import datetime
from monitor import monitor


def job():
    """Job que será executado periodicamente"""
    print(f"\n{'='*60}")
    print(f"Execução iniciada em: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print(f"{'='*60}")
    
    try:
        monitor()
    except Exception as e:
        print(f"✗ Erro durante execução: {e}")
    
    print(f"\nPróxima execução em 1 hora...")


def main():
    """Função principal do scheduler"""
    print("🚀 T CrB Monitoring Scheduler")
    print("⏰ Intervalo: A cada hora")
    print(f"📅 Iniciado em: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print("\nPressione Ctrl+C para parar\n")
    
    # Agenda execução a cada hora
    schedule.every().hour.do(job)
    
    # Executa imediatamente na primeira vez
    print("Executando primeira coleta...")
    job()
    
    # Loop principal
    try:
        while True:
            schedule.run_pending()
            time.sleep(60)  # Verifica a cada minuto
    except KeyboardInterrupt:
        print("\n\n🛑 Scheduler encerrado pelo usuário")
        return 0
    except Exception as e:
        print(f"\n✗ Erro no scheduler: {e}")
        return 1


if __name__ == "__main__":
    sys.exit(main())